cdap-source-control/src/main/java/io/cdap/cdap/sourcecontrol/operationrunner/InMemorySourceControlOperationRunner.java

```diff
@@ -0,0 +1,134 @@
+/*
+ * Copyright © 2023 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package io.cdap.cdap.sourcecontrol.operationrunner;
+
+import com.google.gson.Gson;
+import com.google.gson.GsonBuilder;
+import com.google.inject.Inject;
+import io.cdap.cdap.proto.ApplicationDetail;
+import io.cdap.cdap.proto.id.NamespaceId;
+import io.cdap.cdap.proto.sourcecontrol.RepositoryConfig;
+import io.cdap.cdap.sourcecontrol.CommitMeta;
+import io.cdap.cdap.sourcecontrol.NoChangesToPushException;
+import io.cdap.cdap.sourcecontrol.RepositoryManager;
+import io.cdap.cdap.sourcecontrol.RepositoryManagerFactory;
+import io.cdap.cdap.sourcecontrol.UnexpectedRepositoryChangesException;
+import org.eclipse.jgit.api.errors.GitAPIException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.FileWriter;
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.Collections;
+
+/**
+ * In-Memory implementation for {@link SourceControlOperationRunner}.
+ * Runs all git operation inside calling service.
+ */
+public class InMemorySourceControlOperationRunner implements SourceControlOperationRunner {
+  private final RepositoryManagerFactory repoManagerFactory;
+  private static final Gson GSON = new GsonBuilder().setPrettyPrinting().create();
+  private static final Logger LOG = LoggerFactory.getLogger(InMemorySourceControlOperationRunner.class);
+
+  @Inject
+  public InMemorySourceControlOperationRunner(RepositoryManagerFactory repoManagerFactory) {
```

Comment:
Ok if the factory is to allow us to use a mock then it's fine. We don't need some complicated guice factory bindings though, just have concrete implementations of the RepositoryManagerFactory. Then somebody can look at all the implementations of the factory and understand it immediately instead of being buried somewhere in some guice binding and somewhere else in a unit test.

But the constructor should still not be public for tests. It's fine to be package private and manually construct it in tests, but it should not be public, that means there are likely some abstraction boundaries being violated. I only see it being created in InMemorySourceControlOperationRunnerTest.java, is there somewhere else?

Reply:
got it. will make `RepositoryManagerFactory` a concrete implementation. Currently only InMemorySourceControlOperationRunnerTest is calling the constructor, but we plan to create it in `RemoteSourceControlOperationRunner` task, which is WIP